diff --git a/setup.py b/setup.py
index 9edd230..6665633 100644
--- a/setup.py
+++ b/setup.py
@@ -22,7 +22,7 @@ for package in sonic_dependencies:
 
 setup(
     name='sonic-platform-common',
-    version='1.0',
+    version='1.0.1',
     description='Platform-specific peripheral hardware interface APIs for SONiC',
     license='Apache 2.0',
     author='SONiC Team',
diff --git a/sonic_platform_base/sonic_sfp/sff8024.py b/sonic_platform_base/sonic_sfp/sff8024.py
index 19f81a3..7a6bdfa 100644
--- a/sonic_platform_base/sonic_sfp/sff8024.py
+++ b/sonic_platform_base/sonic_sfp/sff8024.py
@@ -274,6 +274,10 @@ sm_media_interface = {
     '3d': '25G-LR-BiDi',
     '3e': '400ZR, DWDM, amplified',
     '3f': '400ZR, Single Wavelength, Unamplified',
+    '46': 'ZR400-OFEC-16QAM',
+    '47': 'ZR300-OFEC-8QAM',
+    '48': 'ZR200-OFEC-QPSK',
+    '49': 'ZR100-OFEC-QPSK',
     }
 
 passive_copper_media_interface = {
diff --git a/sonic_platform_base/sonic_xcvr/api/credo/__init__.py b/sonic_platform_base/sonic_xcvr/api/credo/__init__.py
new file mode 100644
index 0000000..e69de29
diff --git a/sonic_platform_base/sonic_xcvr/api/public/c_cmis.py b/sonic_platform_base/sonic_xcvr/api/public/c_cmis.py
index bb7e3a7..03f76bd 100644
--- a/sonic_platform_base/sonic_xcvr/api/public/c_cmis.py
+++ b/sonic_platform_base/sonic_xcvr/api/public/c_cmis.py
@@ -487,6 +487,11 @@ class CCmisApi(CmisApi):
         ========================================================================
         """
         trans_dom_th = super(CCmisApi,self).get_transceiver_threshold_info()
+
+        # TODO: Xcvrd doesn't use the vdm alarms.
+        #       When xcvrd implements these alarms, restore here.
+        return trans_dom_th
+
         try:
             trans_dom_th['biasxihighalarm'] = self.vdm_dict['Modulator Bias X/I [%]'][1][1]
             trans_dom_th['biasxilowalarm'] = self.vdm_dict['Modulator Bias X/I [%]'][1][2]
diff --git a/sonic_platform_base/sonic_xcvr/api/public/cmis.py b/sonic_platform_base/sonic_xcvr/api/public/cmis.py
index a677f32..3c32297 100644
--- a/sonic_platform_base/sonic_xcvr/api/public/cmis.py
+++ b/sonic_platform_base/sonic_xcvr/api/public/cmis.py
@@ -282,6 +282,11 @@ class CmisApi(XcvrApi):
             "txbiaslowwarning": float("{:.3f}".format(thresh[consts.TX_BIAS_LOW_WARNING_FIELD]*tx_bias_scale))
             if tx_bias_scale is not None else 'N/A'
         }
+
+        # TODO: Xcvrd doesn't use the alarms of laser_temp, post_fec_ber, pre_fec_ber and vdm info.
+        #       When xcvrd implements these alarms, restore here.
+        return threshold_info_dict
+
         laser_temp_dict = self.get_laser_temperature()
         try:
             threshold_info_dict['lasertemphighalarm'] = laser_temp_dict['high alarm']
@@ -2019,7 +2024,7 @@ class CmisApi(XcvrApi):
             key = "{}_{}".format(consts.HOST_ELECTRICAL_INTERFACE, app)
             val = dic.get(key)
             if val in [None, 'Unknown', 'Undefined']:
-                break
+                continue
             buf['host_electrical_interface_id'] = val
 
             prefix = map.get(self.xcvr_eeprom.read(consts.MEDIA_TYPE_FIELD))
diff --git a/sonic_platform_base/sonic_xcvr/codes/credo/__init__.py b/sonic_platform_base/sonic_xcvr/codes/credo/__init__.py
new file mode 100644
index 0000000..e69de29
diff --git a/sonic_platform_base/sonic_xcvr/codes/public/cmis.py b/sonic_platform_base/sonic_xcvr/codes/public/cmis.py
index b757635..108fdca 100644
--- a/sonic_platform_base/sonic_xcvr/codes/public/cmis.py
+++ b/sonic_platform_base/sonic_xcvr/codes/public/cmis.py
@@ -72,50 +72,50 @@ class CmisCodes(Sff8024):
 
     VDM_TYPE = {
         # VDM_ID: [VDM_NAME, DATA_TYPE, SCALE]
-        1: ['Laser Age [%]', 'U16', 1],
-        2: ['TEC Current [%]', 'S16', 100.0/32767],
-        3: ['Laser Frequency Error [MHz]', 'S16', 10],
-        4: ['Laser Temperature [C]', 'S16', 1.0/256],
-        5: ['eSNR Media Input [dB]', 'U16', 1.0/256],
-        6: ['eSNR Host Input [dB]', 'U16', 1.0/256],
-        7: ['PAM4 Level Transition Parameter Media Input [dB]', 'U16', 1.0/256],
-        8: ['PAM4 Level Transition Parameter Host Input [dB]', 'U16', 1.0/256],
-        9: ['Pre-FEC BER Minimum Media Input', 'F16', 1],
-        10: ['Pre-FEC BER Minimum Host Input', 'F16', 1],
-        11: ['Pre-FEC BER Maximum Media Input', 'F16', 1],
-        12: ['Pre-FEC BER Maximum Host Input', 'F16', 1],
+        # 1: ['Laser Age [%]', 'U16', 1],
+        # 2: ['TEC Current [%]', 'S16', 100.0/32767],
+        # 3: ['Laser Frequency Error [MHz]', 'S16', 10],
+        # 4: ['Laser Temperature [C]', 'S16', 1.0/256],
+        # 5: ['eSNR Media Input [dB]', 'U16', 1.0/256],
+        # 6: ['eSNR Host Input [dB]', 'U16', 1.0/256],
+        # 7: ['PAM4 Level Transition Parameter Media Input [dB]', 'U16', 1.0/256],
+        # 8: ['PAM4 Level Transition Parameter Host Input [dB]', 'U16', 1.0/256],
+        # 9: ['Pre-FEC BER Minimum Media Input', 'F16', 1],
+        # 10: ['Pre-FEC BER Minimum Host Input', 'F16', 1],
+        # 11: ['Pre-FEC BER Maximum Media Input', 'F16', 1],
+        # 12: ['Pre-FEC BER Maximum Host Input', 'F16', 1],
         13: ['Pre-FEC BER Average Media Input', 'F16', 1],
-        14: ['Pre-FEC BER Average Host Input', 'F16', 1],
-        15: ['Pre-FEC BER Current Value Media Input', 'F16', 1],
-        16: ['Pre-FEC BER Current Value Host Input', 'F16', 1],
-        17: ['Errored Frames Minimum Media Input', 'F16', 1],
-        18: ['Errored Frames Minimum Host Input', 'F16', 1],
-        19: ['Errored Frames Maximum Media Input', 'F16', 1],
-        20: ['Errored Frames Maximum Host Input', 'F16', 1],
-        21: ['Errored Frames Average Media Input', 'F16', 1],
-        22: ['Errored Frames Average Host Input', 'F16', 1],
-        23: ['Errored Frames Current Value Media Input', 'F16', 1],
-        24: ['Errored Frames Current Value Host Input', 'F16', 1],
-        128: ['Modulator Bias X/I [%]', 'U16', 100.0/65535],
-        129: ['Modulator Bias X/Q [%]', 'U16', 100.0/65535],
-        130: ['Modulator Bias Y/I [%]', 'U16', 100.0/65535],
-        131: ['Modulator Bias Y/Q [%]', 'U16', 100.0/65535],
-        132: ['Modulator Bias X_Phase [%]', 'U16', 100.0/65535],
-        133: ['Modulator Bias Y_Phase [%]', 'U16', 100.0/65535],
-        134: ['CD high granularity, short link [ps/nm]', 'S16', 1],
-        135: ['CD low granularity, long link [ps/nm]', 'S16', 20],
-        136: ['DGD [ps]', 'U16', 0.01],
-        137: ['SOPMD [ps^2]', 'U16', 0.01],
-        138: ['PDL [dB]', 'U16', 0.1],
-        139: ['OSNR [dB]', 'U16', 0.1],
-        140: ['eSNR [dB]', 'U16', 0.1],
-        141: ['CFO [MHz]', 'S16', 1],
-        142: ['EVM_modem [%]', 'U16', 100.0/65535],
+        # 14: ['Pre-FEC BER Average Host Input', 'F16', 1],
+        # 15: ['Pre-FEC BER Current Value Media Input', 'F16', 1],
+        # 16: ['Pre-FEC BER Current Value Host Input', 'F16', 1],
+        # 17: ['Errored Frames Minimum Media Input', 'F16', 1],
+        # 18: ['Errored Frames Minimum Host Input', 'F16', 1],
+        # 19: ['Errored Frames Maximum Media Input', 'F16', 1],
+        # 20: ['Errored Frames Maximum Host Input', 'F16', 1],
+        # 21: ['Errored Frames Average Media Input', 'F16', 1],
+        # 22: ['Errored Frames Average Host Input', 'F16', 1],
+        # 23: ['Errored Frames Current Value Media Input', 'F16', 1],
+        # 24: ['Errored Frames Current Value Host Input', 'F16', 1],
+        # 128: ['Modulator Bias X/I [%]', 'U16', 100.0/65535],
+        # 129: ['Modulator Bias X/Q [%]', 'U16', 100.0/65535],
+        # 130: ['Modulator Bias Y/I [%]', 'U16', 100.0/65535],
+        # 131: ['Modulator Bias Y/Q [%]', 'U16', 100.0/65535],
+        # 132: ['Modulator Bias X_Phase [%]', 'U16', 100.0/65535],
+        # 133: ['Modulator Bias Y_Phase [%]', 'U16', 100.0/65535],
+        # 134: ['CD high granularity, short link [ps/nm]', 'S16', 1],
+        # 135: ['CD low granularity, long link [ps/nm]', 'S16', 20],
+        # 136: ['DGD [ps]', 'U16', 0.01],
+        # 137: ['SOPMD [ps^2]', 'U16', 0.01],
+        # 138: ['PDL [dB]', 'U16', 0.1],
+        # 139: ['OSNR [dB]', 'U16', 0.1],
+        # 140: ['eSNR [dB]', 'U16', 0.1],
+        # 141: ['CFO [MHz]', 'S16', 1],
+        # 142: ['EVM_modem [%]', 'U16', 100.0/65535],
         143: ['Tx Power [dBm]', 'S16', 0.01],
-        144: ['Rx Total Power [dBm]', 'S16', 0.01],
-        145: ['Rx Signal Power [dBm]', 'S16', 0.01],
-        146: ['SOP ROC [krad/s]', 'U16', 1],
-        147: ['MER [dB]', 'U16', 0.1]
+        # 144: ['Rx Total Power [dBm]', 'S16', 0.01],
+        # 145: ['Rx Signal Power [dBm]', 'S16', 0.01],
+        # 146: ['SOP ROC [krad/s]', 'U16', 1],
+        # 147: ['MER [dB]', 'U16', 0.1]
     }
 
     CDB_FAIL_STATUS = {
diff --git a/sonic_platform_base/sonic_xcvr/mem_maps/credo/__init__.py b/sonic_platform_base/sonic_xcvr/mem_maps/credo/__init__.py
new file mode 100644
index 0000000..e69de29
